### ISSUE TITLE
Finish implementation of ACIA handler

### ISSUE THREAD

**Issue by DavidGZ** [freemint/freemint#177]
As a follow up from a side discussion started in this issue [1].

To finish the implementation of MiNT's own ACIA handler all handlers for vectors in the kbdvecs table must be implemented, just calling the ROM for the missing ones doesn't work. midisys vector handler is the one missing, it could be that there are others I haven't looked deeper.

A bonus of finishing this implementation is that we could probably get rid of the FireTOS workaround for the keyboard handling.

[1] https://github.com/freemint/freemint/issues/164

**Comment by vinriviere** [comment]
Why calling the ROM for the missing ones doesn't work?

Note: non-Atari hardware does not use the ACIA vector (as there is no ACIA at all). kbdvec, mousevec and joyvec are called directly by EmuTOS (from EmuTOS ikbdsys or other function). It must be ensured that it will still work.

**Comment by DavidGZ** [comment]
> Why calling the ROM for the missing ones doesn't work?
> 

I explained it here [1], but I've seen the code from some TOS BIOS sources which are around, I think I got them from dev-docs, and my explanation doesn't fit. I think to remember I arrived to that conclusion after debugging with Hatari, I'll need to do it again because now I have doubts.

Previously when MiNT's ACIA handler was hooked calling midisys crash the system, and I think this is the reason why calling midisys from that handler was commented out.

[1] https://github.com/freemint/freemint/issues/164#issuecomment-624519995

**Comment by DavidGZ** [comment]
Calling the ROM doesn't work with TOS 1, with TOS 2 is OK, and I guess TOS 3 and 4 are also Ok although I didn't check. My affirmation above came because I saw this while testing TOS 1.

This is TOS 1 ACIA handler, look that clears a5 before calling midisys and ikbdsys

```
$00e03ca0 : 48e7 f0f4                          movem.l   d0-d3/a0-a3/a5,-(sp)
$00e03ca4 : 9bcd                               suba.l    a5,a5   <-------------------
$00e03ca6 : 246d 0e8a                          movea.l   $e8a(a5),a2
$00e03caa : 4e92                               jsr       (a2)  <------- jump to midisys
$00e03cac : 246d 0e8e                          movea.l   $e8e(a5),a2
$00e03cb0 : 4e92                               jsr       (a2)
$00e03cb2 : 082d 0004 fa01                     btst      #4,$fa01(a5)
$00e03cb8 : 67ec                               beq.s     $e03ca6
$00e03cba : 08ad 0006 fa11                     bclr      #6,$fa11(a5)
$00e03cc0 : 4cdf 2f0f                          movem.l   (sp)+,d0-d3/a0-a3/a5
$00e03cc4 : 4e73                               rte       
```

This is midisys handler, if we call it from MiNT's ACIA handler a5 will have a bogus value and system crashes. TOS 2 handlers are different and doesn't do anything with a5.

```
$00e03cc6 : 41ed 0de0                          lea       $de0(a5),a0
$00e03cca : 43ed fc04                          lea       $fc04(a5),a1
$00e03cce : 246d 0e76                          movea.l   $e76(a5),a2
$00e03cd2 : 600e                               bra.s     $e03ce2
$00e03cd4 : 41ed 0cd2                          lea       $cd2(a5),a0
$00e03cd8 : 43f9 ffff fc00                     lea       $fffffc00,a1
$00e03cde : 246d 0e72                          movea.l   $e72(a5),a2
$00e03ce2 : 1429 0000                          move.b    0(a1),d2
$00e03ce6 : 0802 0007                          btst      #7,d2
$00e03cea : 671c                               beq.s     $e03d08
$00e03cec : 0802 0000                          btst      #0,d2
$00e03cf0 : 670a                               beq.s     $e03cfc
$00e03cf2 : 48e7 20e0                          movem.l   d2/a0-a2,-(sp)
$00e03cf6 : 6112                               bsr.s     $e03d0a
$00e03cf8 : 4cdf 0704                          movem.l   (sp)+,d2/a0-a2
$00e03cfc : 0202 0020                          andi.b    #$20,d2
$00e03d00 : 6706                               beq.s     $e03d08
$00e03d02 : 1029 0002                          move.b    2(a1),d0
$00e03d06 : 4ed2                               jmp       (a2)
$00e03d08 : 4e75                               rts       

```
If I clear a5 from MiNT's ACIA handler system doesn't crash anymore in the same way but still hangs in some kind of infinitive loop, this also happen with TOS 2. Then I tested saving a5 in the stack too when entering the ACIA handler, and everything seems to work ok. So it seems although we're only saving address registers from a0-a3, the code later relies on a5 not being touch.

So calling the ROM seems OK for TOS 1 if we do some workarounds in the ACIA handler. Still we would need to test other TOS (TOS 3, TOS 4, Milan's TOS)

**Comment by th-otto** [comment]
On Donnerstag, 14. Mai 2020 09:02:17 CEST DavidGZ wrote:
> This is TOS 1 ACIA handler, look that clears a5 before calling midisys and
> ikbdsys

Yes, thats done in a lot of places in TOS 1.x. Because the original stupid 
as68 could not correctly assemble (xxx).w, a5 was used there to address system 
variables and I/O addresses. For that to work, it must obviously be zero.

Generally, i agree with David, that it is much easier to just replace the 
whole ACIA handler, rather than doing lots of workarounds for various TOS 
versions. Only exception seems to be needed when running EmuTOS on non-atari-
hardware, when no ACIA is there; in that case we should rely on EmuTOS to do 
the right thing.

**Comment by vinriviere** [comment]
Good investigation, @DavidGZ!
Really, this demonstrates that the IKBD vectors are designed to be hooked, and not to be called by anything else than TOS.

I'm a bit lost, so let's do a summary of our goal:

1.  FreeMiNT wants keyboard scancodes. So preferably, it hooks the kbdvec vector, as this is the normal way. Unfortunately, it is only available in TOS >= 2 and EmuTOS (Atari or non-Atari hardware). And it does not work as expected with FireTOS (problem with dead keys in issue #164)
2. Alternative for problematic systems would be to hook ikbdsys to deal with the IKBD ACIA. Does this work on TOS 1? And on FireTOS?
3. Last alternative is to hook the whole ACIA vector at 0x118. But it needs to care about both IKBD and MIDI ACIAs.

But definitely, FreeMiNT should hook vectors as less of possible, thus preferring solutions in the above order.

Is the above correct, or did I miss something?

**Comment by DavidGZ** [comment]
> Is the above correct, or did I miss something?

Yes, it's correct. I opened this issue as a Thorsten request to continue a side discussion on issue #164. Should we implement in MiNT our keyboard and midi drivers? I've just wanted to document here, now that the subject is hot in my head, the issues of doing this and the benefits, so if in the future someone picks up again this again all the previous work is here gathered, even for myself that I'll forget almost everything around this in the next weeks ;-)
For now everything seems to work and the kernel has taken care of all TOS cases, we're not in a hurry to fix anything, so it's only a matter of what the kernel should do, what is more appropriate and clean.

**Comment by mikrosk** [comment]
> But definitely, FreeMiNT should hook vectors as less of possible, thus preferring solutions in the above order.

Actually, this is up for debate - one could argue that FreeMiNT should be as standalone as possible, i.e. following Linux, NetBSD etc. On the other hand, we surely want TOS compatibility, so it should mimic its functionality.

That ikbd vectors are somewhere in-between (X)BIOS and GEMDOS, so it is hard to find a definitive answer whether they should or shouldn't be replaced.

**Comment by vinriviere** [comment]
> Actually, this is up for debate - one could argue that FreeMiNT should be as standalone as possible, i.e. following Linux, NetBSD etc. On the other hand, we surely want TOS compatibility, so it should mimic its functionality.

EMUTOS.PRG does exactly that. Completely kick out the previous OS from RAM (TOS, or even AmigaOS 😄) and mimic TOS behaviour by itself. FreeMiNT could certainly do the same. Currently, I see FreeMiNT over TOS like Windows 95 over MS-DOS. It could indeed evolve to become completely independent from the underlying BIOS, providing its own drivers for everything. That's for sure a wide debate. It would be worth to discuss that on the mailing list, or open a separate issue.

**Comment by skaftetryne** [comment]
> On the other hand, we surely want TOS compatibility, so it should mimic its functionality.

Absolutely. TOS compatibility must have top priority. The reason is simple - virtually no applications are developed or maintained anymore. So we must be able to use those 20-30 year old applications, even when they do things that's not really appropriate in a modern world.
